node/service/src/tests.rs

```diff
@@ -0,0 +1,715 @@
+	}
+
+	pub fn add_block<'a>(&'a mut self, hash: Hash, parent_hash: Hash, number: u32) -> &'a mut Self {
+		assert!(number != 0, "cannot add duplicate genesis block");
+		assert!(hash != Self::GENESIS_HASH, "cannot add block with genesis hash");
+		assert!(parent_hash != Self::GENESIS_PARENT_HASH, "cannot add block with genesis parent hash");
+		assert!(self.0.blocks_by_hash.len() < u8::MAX.into());
+		self.add_block_inner(hash, parent_hash, number)
+	}
+
+	fn add_block_inner<'a>(&'a mut self, hash: Hash, parent_hash: Hash, number: u32) -> &'a mut Self {
+		let header = ChainBuilder::make_header(parent_hash, number);
+		assert!(self.0.blocks_by_hash.insert(hash, header).is_none(), "block with hash {:?} already exists", hash,);
+		self.0.blocks_at_height.entry(number).or_insert_with(Vec::new).push(hash);
+		self
+	}
+
+	pub fn ff_approved(
+		&mut self,
+		branch_tag: u8,
+		parent: Hash,
+		block_number: BlockNumber,
+	) -> Hash {
+		let block = self.ff(branch_tag, parent, block_number);
+		let _ = self.0.approved_blocks.insert(block);
+		block
+	}
+
+	/// Add a relay chain block that contains a disputed parachain block.
+	/// For simplicity this is not modeled explicitly.
+	pub fn ff_disputed(
+		&mut self,
+		branch_tag: u8,
+		parent: Hash,
+		block_number: BlockNumber,
+	) -> Hash {
+		let block = self.ff_approved(branch_tag, parent, block_number);
+		let _ = self.0.disputed_blocks.insert(block);
+		block
+	}
+
+	pub fn ff(&mut self, branch_tag: u8, parent: Hash, block_number: BlockNumber) -> Hash {
+		let hash = Hash::repeat_byte((block_number as u8 | branch_tag) as u8);
+		let _ = self.add_block(hash, parent, block_number);
+		hash
+	}
+
+	pub fn set_heads(&mut self, heads: impl IntoIterator<Item = Hash>) {
+		self.0.heads = heads.into_iter().collect();
+	}
+
+	pub fn init(self) -> TestChainStorage {
+		self.0
+	}
+
+	fn make_header(parent_hash: Hash, number: u32) -> Header {
+		let digest = {
+			let mut digest = Digest::default();
+			let (vrf_output, vrf_proof) = garbage_vrf();
+			digest.push(DigestItem::babe_pre_digest(PreDigest::SecondaryVRF(SecondaryVRFPreDigest {
+				authority_index: 0,
+				slot: 1.into(), // slot, unused
+				vrf_output,
+				vrf_proof,
+			})));
+			digest
+		};
+
+		Header { digest, extrinsics_root: Default::default(), number, state_root: Default::default(), parent_hash }
+	}
+}
+
+/// Generalized sequence of the test, based on
+/// the messages being sent out by the `fn finality_target`
+/// Depends on a particular `target_hash`
+/// that is passed to `finality_target` block number.
+async fn test_skeleton(
+	chain: &TestChainStorage,
+	virtual_overseer: &mut VirtualOverseer,
+	target_block_hash: Hash,
+	best_chain_containing_block: Option<Hash>,
+	highest_approved_ancestor_block: Option<HighestApprovedAncestorBlock>,
+	undisputed_chain: Option<Hash>,
+) {
+	let undisputed_chain = undisputed_chain.map(|x| (chain.number(x).unwrap().unwrap(), x));
+
+	println!("best leaf response: {:?}", undisputed_chain);
```

Comment:
nit: I see a mix of `dbg!`, `println` and `tracing::trace!`, we probably want to leave only the latter?

Reply:
`'static` is required in this position. I did not investigate where the requiremt came from.